src/LinAlg/hiopVectorCuda.hpp

```diff
@@ -0,0 +1,342 @@
+  virtual bool matchesPattern(const hiopVector& ix);
+  /// @brief dual adjustment -> see hiopIterate::adjustDuals_primalLogHessian
+  virtual void adjustDuals_plh(const hiopVector& x, const hiopVector& ix, const double& mu, const double& kappa);
+
+  /// @brief True if all elements of this are zero. TODO: add unit test
+  virtual bool is_zero() const;
+  /// @brief check for nans in the local vector
+  virtual bool isnan_local() const;
+  /// @brief check for infs in the local vector
+  virtual bool isinf_local() const;
+  /// @brief check if all values are finite /well-defined floats. Returns false if nan or infs are present.
+  virtual bool isfinite_local() const;
+
+  /// @brief prints up to max_elems (by default all), on rank 'rank' (by default on all)
+  virtual void print(FILE* file=nullptr, const char* message=nullptr,int max_elems=-1, int rank=-1) const;
+  /// @brief allocates a vector that mirrors this, but doesn't copy the values
+  virtual hiopVector* alloc_clone() const;
+  /// @brief allocates a vector that mirrors this, and copies the values
+  virtual hiopVector* new_copy () const;
+
+  /* more accessers */
+  inline size_type get_local_size() const { return n_local_; }
+  inline double* local_data() { return data_dev_; }
+  inline const double* local_data_const() const { return data_dev_; }
+  inline double* local_data_host() { return data_host_; }
+  inline const double* local_data_host_const() const { return data_host_; }
+
+  virtual void copyToDev();
+  virtual void copyFromDev();
+  virtual void copyToDev() const;
+  virtual void copyFromDev() const;
+
+  /// @brief get number of values that are less than the given value 'val'. TODO: add unit test
+  virtual size_type numOfElemsLessThan(const double &val) const;
+  /// @brief get number of values whose absolute value are less than the given value 'val'. TODO: add unit test
+  virtual size_type numOfElemsAbsLessThan(const double &val) const;  
+
+  /// @brief set int array 'arr', starting at `start` and ending at `end`, to the values in `arr_src` from 'start_src`
+  /// TODO: add unit test
+  virtual void set_array_from_to(hiopInterfaceBase::NonlinearityType* arr, 
+                                 const int start, 
+                                 const int end, 
+                                 const hiopInterfaceBase::NonlinearityType* arr_src,
+                                 const int start_src) const;
+  virtual void set_array_from_to(hiopInterfaceBase::NonlinearityType* arr, 
+                                 const int start, 
+                                 const int end, 
+                                 const hiopInterfaceBase::NonlinearityType arr_src) const;
+
+
+
+
+  /// @brief check if `this` vector is identical to `vec`
+  virtual bool is_equal(const hiopVector& vec) const;
+
+  /* functions for this class */
+  inline MPI_Comm get_mpi_comm() const { return comm_; }
+
+private:
+  MPI_Comm comm_;
+  double* data_host_;
+  double* data_dev_;
```

Comment:
Given the implementation, consider using more intuitive variable names:
```suggestion
  double* data_host_mirror_;
  double* data_;
```
Your data is processed on device. You copy it to the host mirror for special purposes (e.g. for printing during debugging).